README.md
Backyard Watering
======

My DIY backyard smart-watering project. (When I have the time to tinker)


Dependencies
------

[fivdi](https://github.com/fivdi)'s [mcp-spi-adc](https://github.com/fivdi/mcp-spi-adc) and [pigpio](https://github.com/fivdi/pigpio)


History.md
Version History
======

## 0.0.5 / Sept 22 2019
* Display percentage value for each sensor reading

## 0.0.4 / Nov 08 2018
* Test opening and closing a solenoid valve for the water line

## 0.0.3 / Oct 15 2018
* Show values of all five sensors and average value

## 0.0.2 / Sep 08 2018
* Temperature sensor and five soil moisture sensors online

## 0.0.1 / Sep 08 2018
* Combined code to read from temperature sensor and one moisture sensor

## 0.0.0 / Feb 27 2018
* Tested moisture sensor code


valve-test.js
const Gpio = require('pigpio').Gpio;

const valve = new Gpio(17, {mode: Gpio.OUTPUT});

var level = 0;

function getDateTime() {

    var date = new Date();

    var hour = date.getHours();
    hour = (hour < 10 ? "0" : "") + hour;

    var min  = date.getMinutes();
    min = (min < 10 ? "0" : "") + min;

    var year = date.getFullYear();

    var month = date.getMonth() + 1;
    month = (month < 10 ? "0" : "") + month;

    var day  = date.getDate();
    day = (day < 10 ? "0" : "") + day;

    return hour + ":" + min + " on " + month + "/" + day + "/" + year;

}

function closeValve() {
    level = 0;
    valve.digitalWrite(0);
}

function openValve() {
    level = 1;
    valve.digitalWrite(1);
}

function toggleValve() {
    if (level == 0) {
        level = 1;
    } else {
        level = 0;
    }
    valve.digitalWrite(level);
}

function getState() {
    if (level == 1) {
        return "open";
    } else {
        return "closed";
    }
}

var curTime = getDateTime();

console.log("Testing solenoid valve opening and closing.\r\n");

closeValve();   // Make sure we start with the valve closed.

setTimeout(function() {
    curTime = getDateTime();
    openValve();
    console.log("Valve was opened at " + curTime + "\r\n");
}, 10000);

setTimeout(function() {
    curTime = getDateTime();
    closeValve();
    console.log("Valve was closed at " + curTime + "\r\n");
}, 60000);


moisture.js
var mcpadc = require('mcp-spi-adc');

var moistureSensor = mcpadc.open(2, {speedHz: 20000}, function (err) {
    if (err) throw err;

    var moistureValues = [];

    setInterval(function () {
        moistureSensor.read(function (err, reading) {
            if (err) throw err;
            var mvoltage = reading.rawValue,
		voltage  = mvoltage / 1000,
                range    = (voltage / 0.8) * 100,    // Using 3.3V; will differ for 5V
		moistMax = 0,
		moistMin = 0,
		moistDiff = 0;
            moistureValues.push(voltage.toFixed(3));
            moistMax = Math.max.apply(null, moistureValues);
            moistMin = Math.min.apply(null, moistureValues);
	    moistDiff = moistMax - moistMin;
            //console.log("Voltage: " + voltage.toFixed(4) + " => " + range.toFixed(1) + "%");
	    console.log("Current: " + voltage.toFixed(3) + "V, Min: " + moistMin + "V, Max: " + moistMax + "V => Difference: " + moistDiff.toFixed(3));
        });
    }, 1000);
});


backyard.js
const factor = 0.855; // Using 3.3V. This will differ for 5V

var mcpadc = require('mcp-spi-adc');
var t1  = 0.0,
    m1  = 0.0,
    m2  = 0.0,
    m3  = 0.0,
    m4  = 0.0,
    m5  = 0.0,
    avg = 0.0;

var tempSensor = mcpadc.open(0, {speedHz: 20000}, function (err) {
    if (err) throw err;

    setInterval(function () {
        tempSensor.read(function (err, reading) {
            if (err) throw err;
            var voltage = reading.value,
                degC    = (reading.value * 3.3 - 0.5) * 100,
                degF    = degC * (9 / 5) + 32;
            t1 = degF.toFixed(1);
        });
    }, 1000);
});

var moistureSensor1 = mcpadc.open(1, {speedHz: 20000}, function (err) {
    if (err) throw err;

    setInterval(function () {
        moistureSensor1.read(function (err, reading) {
            if (err) throw err;
            m1 = reading.value;
        });
    }, 1000);
});


// No plant / no drip line nearby: soil will be very dry, causing skewed average
var moistureSensor2 = mcpadc.open(2, {speedHz: 20000}, function (err) {
    if (err) throw err;

    setInterval(function () {
        moistureSensor2.read(function (err, reading) {
            if (err) throw err;
            m2 = reading.value;
        });
    }, 1000);
});


var moistureSensor3 = mcpadc.open(3, {speedHz: 20000}, function (err) {
    if (err) throw err;

    setInterval(function () {
        moistureSensor3.read(function (err, reading) {
            if (err) throw err;
            m3 = reading.value;
        });
    }, 1000);
});

var moistureSensor4 = mcpadc.open(4, {speedHz: 20000}, function (err) {
    if (err) throw err;

    setInterval(function () {
        moistureSensor4.read(function (err, reading) {
            if (err) throw err;
            m4 = reading.value;
        });
    }, 1000);
});

var moistureSensor5 = mcpadc.open(5, {speedHz: 20000}, function (err) {
    if (err) throw err;

    setInterval(function () {
        moistureSensor5.read(function (err, reading) {
            if (err) throw err;
            m5 = reading.value;
        });
    }, 1000);
});


function getDateTime() {

    var date = new Date();

    var hour = date.getHours();
    hour = (hour < 10 ? "0" : "") + hour;

    var min  = date.getMinutes();
    min = (min < 10 ? "0" : "") + min;

    var year = date.getFullYear();

    var month = date.getMonth() + 1;
    month = (month < 10 ? "0" : "") + month;

    var day  = date.getDate();
    day = (day < 10 ? "0" : "") + day;

    return hour + ":" + min + " on " + month + "/" + day + "/" + year;
}

function getPercentageFromReading(value) {
    var percentage = (value / 0.8) * 100;
    return percentage.toFixed(1);
}

function readoutToConsole() {
    // Clean up some numbers for display
    // f : "fixed"
    // p : "percent"
    var m1f = m1.toFixed(3);
    var m1p = getPercentageFromReading(m1f);
    var m2f = m2.toFixed(3);
    var m2p = getPercentageFromReading(m2f);
    var m3f = m3.toFixed(3);
    var m3p = getPercentageFromReading(m3f);
    var m4f = m4.toFixed(3);
    var m4p = getPercentageFromReading(m4f);
    var m5f = m5.toFixed(3);
    var m5p = getPercentageFromReading(m5f);

    // Get the average of the five moisture sensors
    var avgf   = (m1 + m2 + m3 + m4 + m5) / 5;
        avgf   = avg.toFixed(3);
    var avgp = getPercentageFromReading(avg);

    var curTime = getDateTime();

    console.log("At " + curTime + ":\r\nAmbient temperature: " + t1 + " deg F\nAverage soil moisture reading: " + avgf + "V (" + avgp + "%)\nSensor 1: " + m1f + "V (" + m1p + "%)\nSensor 2: " + m2f + "V (" + m2p +"%)\nSensor 3: " + m3f + "V (" + m3p +"%)\nSensor 4: " + m4f + "V (" + m4p +"%)\nSensor 5: " + m5f + "V (" + m5p +"%)\r\n");
}


var curTime = getDateTime();
console.log("Monitoring backyard watering system.\r\nStarted at " + curTime + "\r\nTaking readings every 5 minutes.\r\n");

// Display some readings without having to wait for the first interval (5 minutes)
// but still allow the sensors to be read (15-second intervals)
setTimeout(function() {
    readoutToConsole();
}, 5000);

setInterval(function() {
    readoutToConsole();
}, 300000);
